contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1705 @@
+
+        if (nodeRight.left == address(0)) {
+            Node storage nodeRightStorage = nodes[C_ND.right];
+
+            nodeRightStorage.left = C_ND.left;
+            nodeRightStorage.sumOfLeftBalances = C_ND.sumOfLeftBalances;
+            nodeRightStorage.height = 1 + Math.max(height(nodeRightStorage.left), height(nodeRightStorage.right));
+
+            delete nodes[_node];
+
+            return C_ND.right;
+        } else {
+            // nodes[_node].balance = 0;
+            // return _node
+
+            Node memory leastMinNode = _findLeastMinNodeForCase1(C_ND.right);
+
+            C_ND.right = _deleteNode(C_ND.right, leastMinNode.node, leastMinNode.balance);
+            delete nodes[_node];
+
+            Node storage lmnStore = nodes[leastMinNode.node];
+
+            // lmn is removed in storage, so create a new one
+            lmnStore.node = leastMinNode.node;
+            lmnStore.balance = leastMinNode.balance;
+            lmnStore.left = C_ND.left;
+            lmnStore.right = C_ND.right;
+            lmnStore.sumOfLeftBalances = C_ND.sumOfLeftBalances;
+
+            Node memory C_ND_right = nodes[C_ND.right];
+            lmnStore.sumOfRightBalances = _getTotalBalancesIncludingWeight(C_ND_right);
+            lmnStore.height = calculateUpdatedHeight(lmnStore);
+
+            return leastMinNode.node;
+        }
+    }
+
+    /// @notice Find the least minimum node for given node
+    /// @param _node Address of node from which least min node has to be found
+    /// @return Copy of Node that will be replaced
+    function _findLeastMinNodeForCase1(address _node) internal view returns (Node memory) {
+        Node memory node = nodes[_node];
+
+        if (node.left != address(0)) {
+            return _findLeastMinNodeForCase1(node.left);
+        }
+
+        return (node);
+    }
+}
+
+contract ClusterSelector is SingleSelector {
+    using ClusterLib for address[];
+    using ClusterLib for bytes;
+
+    constructor() SingleSelector() {}
+
+    /// @notice Select top N clusters
+    /// @return List of addresses selected
+    function selectTopNClusters(uint256 randomizer, uint256 N) public view returns (address[] memory) {
+        require(N <= totalElements, ClusterLib.INSUFFICIENT_ELEMENTS_IN_TREE);
+
+        address[] memory _emptyPath = new address[](0);
+        address[] memory selectedNodes = new address[](N);
+        bytes[] memory pathToSelectedNodes = new bytes[](N);
+
+        for (uint256 index = 0; index < N; index++) {
+            pathToSelectedNodes[index] = abi.encode(_emptyPath);
+        }
+
+        for (uint256 index = 0; index < N; index++) {
+            randomizer = uint256(keccak256(abi.encode(randomizer, index)));
+            Node memory _root = nodes[root];
+            uint256 totalWeightInTree = _getTotalBalancesIncludingWeight(_root);
+            uint256 _sumOfBalancesOfSelectedNodes = sumOfBalancesOfSelectedNodes(selectedNodes);
```

Comment:
Instead of calculating this in every iteration, sum it up over iterations as we already will know the `_node`